docs/configuration.md
# Configuration

## Without a Configuration File

The core configuration can be supplied by environment variables, everything else will default, either to fixed values built into Updates2MQTT, or in the case of the node name, taken from the operating system.

| Env Var       | Default      |
|---------------|--------------|
| MQTT_HOST     | localhost    |
| MQTT_PORT     | 1883         |
| MQTT_USER     | *NO DEFAULT* |
| MQTT_PASSWORD | *NO DEFAULT* |
| MQTT_VERSION  | 3.11.        |
| U2M_LOG_LEVEL | INFO         |

Startup will fail if `MQTT_USER` and `MQTT_PASSWORD` are not defined some how.

The example [docker-compose.yaml](examples/docker_compose.md) and [.env](examples/env.md) demonstrate one way of doing this, or skip
the `.env` file and use an `environment` section in the Compose file.

Set `U2M_AUTOGEN_CONFIG=0` in the environment to prevent a default config file being created in the local compose directory if you want to keep it zero-configuration-file.

## With A Configuration File

The configuration file is optional, and only needed if you have to override the defaults.

Create file `config.yaml` in `conf` directory. If the file is not present, a default file will be generated, and the parent director if necessary. If you don't want that to happen, then set `U2M_AUTOGEN_CONFIG=0`.

### Example configuration file

This is a maximal config file, the minimum is no config file at all, which will generate a default config file. The only mandatory values are the MQTT user name and password, everything else can be omitted ( although
its best to have at least a `node` `name` value so HomeAssistant doesn't show some ugly generated Docker host name).

```yaml title="config.yaml snippet"

node:
  name: docker-host-1 # Unique name for this instance, used to name MQTT entities. Defaults to O/S hostname
  git_repo_path: /usr/bin/git # Path to git inside container, needed only if non-default and using local docker builds
  healthcheck:
    enabled: true
    interval: 300 # publish a heartbeat every 5 minutes
    topic_template: healthcheck/{node_name}/updates2mqtt
mqtt:
  host: ${oc.env:MQTT_HOST}
  user: ${oc.env:MQTT_USER}
  password: ${oc.env:MQTT_PASS}$ # Use an environment variable for secrets
  port: ${oc.env:MQTT_PORT}
  topic_root: updates2mqtt
  protocol: 3.11 # Can be changed to 5 if your broker supports it
homeassistant:
  discovery:
    prefix: homeassistant # Matches the default MQTT discovery prefix in Home Assistant
    enabled: true
  state_topic_suffix: state
  device_creation: true # create a Home Assistant device and associate it with each update entity
  area: "Server Room" # suggest an area for Home Assistant to give to the Device
  extra_attributes: true # use `json_attributes_topic` feature in Home Assistant to source state attributes from discover
  force_command_creation: # create a command topic even if there's no support for automated update
docker:
  enabled: true
  allow_pull: true # if true, will do a `docker pull` if an update is available
  allow_restart: true # if true, will do a `docker-compose up` if an update is installed
  allow_build: true # if true, will do a `docker-compose build` if a git repo is configured
  compose_version: v2 # Controls whether to use `docker-compose` (v1) or `docker compose` (v2) command
  default_entity_picture_url: https://www.docker.com/wp-content/uploads/2022/03/Moby-logo.png # Picture for update dialog
  device_icon: mdi:docker # Material Design Icon to use when browsing entities in Home Assistant
  default_api_backoff: 600 # Default time to back off container registry APIs
  # device_icon: mdi:train-car-container # Alternative icon if you don't like Docker branding
  discover_metadata:
    linuxserver.io:
      enabled: true
      cache_ttl: 604800 # cache metadata for 1 week
  image_ref_select: # limit the containers which will be published to Home Assistant
    include:
    - .*special-dev
    exclude:
    - .*dev
    - .*nightly
  version_select:  # limit the containers which will be published to Home Assistant
     exclude:
     -  .*-beta.*
     include: 
     - 1\..*
scan_interval: 10800 # sleep interval between scan runs, in seconds
log:
  level: INFO
```

## Improving Security

### Moving Secrets Out of Config

Example use of environment variables, e.g. for secrets:

```yaml title="config.yaml snippet"
mqtt:
    password: ${oc.env:MQTT_PASS}
```

### Running as non-root

It is good practice not to run Docker containers as root, and `updates2mqtt` will
work with any user so long as it has Docker permissions, usually as a result
of being a member of the `docker` group.

To create a suitable use, use the shell command below - it will create a user
that can only be used for this purpose, and can't otherwise login. It assumes there is already a group called `docker` with access to the Docker Daemon, if you dont
have one, follow the [Docker Post Install Steps](https://docs.docker.com/engine/install/linux-postinstall/) which explain how and why to do it.

```bash
sudo adduser --system --ingroup docker --no-create-home -shell /sbin/nologin updates2mqtt
```

Note the `uid` that is reported here. If you don't know the `gid` for the `docker` group, use `grep docker /etc/group`. In this example, our `uid` is `130` and the `gid` of `docker` group is `119`.

In the `docker-compose.yaml`, set the user and group using [user](https://docs.docker.com/reference/compose-file/services/#user) attribute:

```yaml title="docker compose snippet"
services:
  updates2mqtt:
    container_name: updates2mqtt
    image: ghcr.io/rhizomatics/updates2mqtt:latest
    user: 130:119
```

If you're using Updates2MQTT to update local git repos, then the user created above will also need `rw` access to those, which you can do by making it a member of the
same group as owns the repos and making sure they have group `rw` access configured.

For more information, see the [Understanding the Docker USER Instruction](https://www.docker.com/blog/understanding-the-docker-user-instruction/) article from Docker.

### MQTT Access Control

Its best to have a dedicated MQTT user for Updates2MQTT, for security and debug. For most
secure installations, only use secure ports with validated certificates, although this will
require more complicated setup and ongoing support, including using host names rather than
IP addresses, and with [LetsEncrypt](https://letsencrypt.org) to update certificates.

The two brokers most commonly used with Home Assistant, **Mosquitto** and **EMQX**, both have
access control mechanisms, so you can restrict the user account for Updates2MQTT to only be able
to read and write its own topics.

## Customizing images and release notes

Individual docker containers can have customized entity pictures or release notes, using env variables, for example in the `docker-compose.yaml` or in a separate `.env` file:

```yaml title="docker compose snippet"
    environment:
      - UPD2MQTT_PICTURE=https://frigate.video/images/logo.svg
      - UPD2MQTT_RELNOTES=https://github.com/blakeblackshear/frigate/releases
```

or using labels

```yaml title="docker compose snippet"
    labels:
      updates2mqtt.picture: https://frigate.video/images/logo.svg
      updates2mqtt.relnotes: https://github.com/blakeblackshear/frigate/releases
```
  l
The images will show up in the *Update* section of *Settings* menu in HomeAssistant,
as will the release notes link. SVG icons should be used.

## Customizing Versions

Updates2MQTT attempts to find the most human-friendly representation of image versions that
can be reliably used. Ideally that's a `v1.5.4` type version (whether formally SemVer or just traditional version).

By default, configurable using `version_policy` in the Docker section of the config, it uses an `auto` version policy that will choose the most meaningful, and fall back to digests where versions aren't available (usually via image labels/annotations). This will also take into account where updates are throttled, or a pinned digest declared in the container.

This can be overridden at container level using using the `updates2mqtt.version_policy` container label or `UPD2MQTT_VERSION_POLICY` environment variable:

   - `AUTO` - to do the best it can with versions, git repo digests, index digests or config digests
   - `VERSION` - always choose simple version unless version not available
     - Some images use version oddly, where its more of a label applying to multiple releases than a version. Also
       there's guarantee for container images that a human friendly version always points to the same thing.
     - This is useful where you know the image has sensible versions and trust it enough
   - `DIGEST` - always use the 12-char abbreviated digest, even if version available
   - `VERSION_DIGEST` - use a `version:1234567890ab` style combo of version and digest id where both available

If the chosen option isn't available, they'll all fail back to `auto`. A diagnostic code, `version-select` that ties back to precisely which [code])(https://github.com/rhizomatics/updates2mqtt/blob/main/src/updates2mqtt/integrations/docker.py#L533) used is included in the attributes.

## Silencing Containers

If there are containers which are changing very frequently with development builds, or for other reasons
shouldn't be published to Home Assistant, then use the `image_ref_select` in configuration.

They will still be published to MQTT but not to the Home Assistant MQTT Discovery topic.

```yaml title="config.yaml snippet"
docker:
  enabled: true
  image_ref_select:
    exclude:
      - .*:nightly
      - .*:dev
```

Alternatively, set `UPD2MQTT_IGNORE` flag on the container itself to completely ignore it.

## API Throttling

Docker API has [usage limits](https://docs.docker.com/docker-hub/usage/) which may be triggered if there are many containers ( and other registries will have similar).

`updates2mqtt` will back off if a `429` Too Many Requests response is received, and pause for that specific registry for the requested number of seconds. There's a default in `docker` config of `default_api_backoff` applied if the backoff can't be automatically determined.

The main technique to avoid throttling is caching of responses, and fortunately many of the calls are cache friendly, such as the manifest retrieval. By default, responses will be cached as suggested by the registry API service ([explanation](https://hishel.com/1.1/specification/#how-it-works)), however this can be overridden with these options:

| Config Key | Default | Comments |
| ---------- | ------- | -------- |
| `mutable_cache_ttl` | None | This is primarily the fetch of `latest` or similar tags to get new versions |
| `immutable_cache_ttl` | 7776000 (90 days) | This is for anything fetched by a digest, such as image manifests. The only limitation for these should be storage space |
| `token_cache_ttl` | None | Caching for authorization tokens, `docker.io` is good for 300 seconds, not all registries publish the life in the response |

The cache, using [Hishel](https://hishel.com), is automatically cleaned up of old entries once the TTL (Time to Live) has expired.

The other approach can be to reduce the scan interval, or ignore some of the containers.

### GitHub API

GitHub REST API has its own throttling, which may impact fetching release summaries. A higher limit can be achieved using a
*personal access token*. Create one in *Developer Settings* in GitHub, make sure it has access to "Public Repositories",
and configure it in Updates2MQTT as below:

```yaml title="updates2mqtt config snippet"
github:
  access_token: my_access_token
```

## Icon Sources

Updates look nicer in Home Assistant with a suitable icon. Updates2mqtt comes
pre-packaged with some common ones, in `common_packages.yaml`, and can automatically fetch them (and release links) for the popular [linuxserver.io](https://www.linuxserver.io) packages.  

If you have something not covered, here are some good places to look for self-hosted app icons:

- [Homarr Dashboard Icons](https://github.com/homarr-labs/dashboard-icons)
- [Self Hosted Icons](https://selfh.st/icons/) ([repo](https://github.com/selfhst/icons))
- [Simple Icons](https://github.com/simple-icons/simple-icons)
- [Tabler Icons](https://tabler.io/icons)
- [Papirus Icons](https://github.com/PapirusDevelopmentTeam/papirus-icon-theme)
- [Homelab SVG Assets](https://github.com/loganmarchione/homelab-svg-assets)

## Environment Variables

The following environment variables can be used to configure containers for `updates2mqtt`:

| Env Var                    | Description                                                                                  | Default         |
|----------------------------|----------------------------------------------------------------------------------------------|-----------------|
| `UPD2MQTT_UPDATE`          | Update mode, either `Passive` or `Auto`. If `Auto`, updates will be installed automatically. | `Passive`       |
| `UPD2MQTT_PICTURE`         | URL to an icon to use in Home Assistant.                                                     | Docker logo URL |
| `UPD2MQTT_RELNOTES`        | URL to release notes for the package.                                                        |                 |
| `UPD2MQTT_GIT_REPO_PATH`   | Relative path to a local git repo if the image is built locally.                             |                 |
| `UPD2MQTT_IGNORE`          | If set to `True`, the container will be ignored by Updates2MQTT.                             | False           |
                        |                 |
| `UPD2MQTT_VERSION_POLICY` | Change how version derived from container label or image hash, `Version`,`Digest`,`Version_Digest` with default of `Auto`|
| `UPD2MQTT_REGISTRY_TOKEN` | Access token for authentication to container distribution API, as alternative to making a call to `token` service |

## Docker Labels

Alternatively, use Docker labels

| Label                          | Env Var                    |
|--------------------------------|----------------------------|
| `updates2mqtt.update`          | `UPD2MQTT_UPDATE`          |
| `updates2mqtt.picture`         | `UPD2MQTT_PCITURE`         |
| `updates2mqtt.relnotes`        | `UPD2MQTT_RELNOTES`        |
| `updates2mqtt.git_repo_path`   | `UPD2MQTT_GIT_REPO_PATH`   |
| `updates2mqtt.ignore`          | `UPD2MQTT_IGNORE`          |
| `updates2mqtt.version_policy`  | `UPD2MQTT_VERSION_POLICY`  |
| `updates2mqtt.registry_token`  | `UPD2MQTT_REGISTRY_TOKEN`  |


```yaml title="Example Compose Snippet"
restarter:
    image: docker:cli
    command: ["/bin/sh", "-c", "while true; do sleep 86400; docker restart mailserver; done"]
    labels:
        updates2mqtt.relnotes: https://component.my.com/release_notes
```

docs/examples/config_maximal.md
# Maximal Configuration

More extensive example configuration, using all the features

```yaml
--8<-- "examples/config.yaml.maximal"
```

CHANGELOG.md
# CHANGELOG

## 1.8.1
### Home Assistant
- Re-publish state every scan to ensure fresh
- Main discovery and state message now better in sync on `in_progress` flag
- State not published during update if Home Assistant mode off
### Version Detection
- Version Policy now includes `TIMESTAMP` for image creation date/time, which is also now part of
auto version algorithm
- Version policy can be also be set in `custom_packages.yaml` or in the `packages` section of config
- Detect common mis-configuration where ubuntu base layer version published as derived image version
### MQTT
- MQTT result code, MID and publish status logged at debug, and at warning if non-success result code
## 1.8.0
### Package Info
- The `common_packages.yaml` that ships inside the Docker image can be extended, or overridden, using the `packages` section of the main config file
- Source platform will be inferred, and verified, for GitHub, Gitlab, Quay, LSCR and Codeberg
### Release Notes
- If the component tag doesn't match the release tag (e.g. `1.7.3` vs `v1.7.3`), an alt lookup will be made on latest github released, and used if the tag name on that release matches the expected tag, with a common `v`,`V`,`r` or `R` prefix
- A GitHub *personal access token* can now be added in the `github:` config section, which will increase number of API calls that can be made for release info before being throttled
- Release notes can be better guessed, subject to URL validation, for Gitlab and GitHub
### CLI
- Logging improved, with different default for container lookup, so there's always some output from commands
- Added a `tags` option to retrieve list of known tags in the registry
### Housekeeping
- MQTT topic clean overhauled - 
    - Now runs only after scan
    - Checks topics against current discoveries vs previous session key based check

## 1.7.3
- Minor fix for stats reporting
## 1.7.2
### API Stats
- OCI v2 API client maintains api stats per host, and errors per status code, with a cache hit ratio and avg elapsed time
### Container Registry Checks
- The Config document is now also pulled in, which means more images have annotations, nice versions and release notes
- `OCI_V2_MINIMAL` option to make only the calls needed for version checks
- Diagnostic code now distinguishes two ways that different digests are checked between old and new
- Stats now tallied for all registry URL calls, and periodically logged
### CLI
- Fixed issues with docker.io fetches
- Colourful and jq friendly JSON output
- Log level now defaults to WARNING to make it easier to pipe results to `jq`, can change to see headers using `log_level=DEBUG`
## 1.7.1
- No code change, version changed to fix a pypi publish issue
## 1.7.0
### py3.14
- Image now built with python 3.14 by default
- Tests update for py3.14 mock async changes
### OCI Container Registry APIs
- By default now uses the OCI v2 Distribution APIs rather than old Docker APIs via Docker SDK
   - Older method can be switched back on using `docker.registry.api: docker_client` configuration
   - Container registry also used where available to pull in annotations ( including 
a meaningful version ) using the OCI Distriubtion API
  - Docker, GitHub GHCR, Gitlab, LSCR, Microsoft MCR and Codeberg configured
  - Authentication logic to automatically adapt and retry dynamically for other platforms
  - Annotations will be sourced from the Index or Manifest, with priority for the (platform-specific) manifest
  - Digest matching adapts to different practices across repos
- Registries can now be selectively included/excluded using `registry_select` in Docker configuration section
### GitHub Release Enrichment
- Where an annotation points to a source URL and hash, this can be used to derive a 
`diff_url`, a direct link to the actual release notes, and a release summary pulled from the
Github source repo releases and passed onto Home Assistant
- A token can be provided, e.g. a Personal Access Token, at container level using `UPD2MQTT_REGISTRY_TOKEN`
- Common packages now have a `source_repo_url` and don't need a release notes URL too if on GitHub
### Version Policies
- Version policies available to make use of meaningful versions from annotations
   - `AUTO` will detect what looks like a SemVer and report only the meaningful version to HA
   - `VERSION`,`DIGEST` and `VERSION_DIGEST` can be used to fix on which fields to use
   - If the chosen option isn't available, next preference is for `version:digest` qualified version, falling back to whatever else is available
   - Version policy can be set at container level using `UPD2MQTT_VERSION_POLICY` env var or corresponding label
- Version selection now has a diagnostic code, `version_basis`, that ties back to precisely which code used
### Local Docker Image Enrichment
- Metadata from local `org.opencontainers` labels published where available
- Added more packages to common-packages, vector and greptime
### Selective Home Assistant Discovery
- New `version_select` option in Docker config works like `image_ref_select` using the reported version instead
### Pinned Digests
- Better handling of tags with a pinned digest, e.g. `ghcr.io/immich-app/postgres:14-vectorchord0.4.3-pgvectors0.2.0@sha256:41eacbe83eca995561fe43814fd4891e16e39632806253848efaf04d3c8a8b84`
### API Throttling
- Throttling fixed for TTL taken from remote API
- Docker provider shuffles the list of containers it scans so if there is aggressive throttling there isn't the same set of containers always getting scanned before the rate limit hits
### Caching
- Index and Manifest local cache age now reported on MQTT message
- Additional logging and defaults for caching
- Local cache TTL override can be removed by using `None` for cache_ttl in registry config

## 1.6.0
### MQTT Topics
- Separated out generic and Home Assistant specific topics
  - The *State*, *Command* and *Discovery* topics are strictly Home Assistant schema
  - The full discovery attributes are now published on their own topic

src/updates2mqtt/config.py
import os
import typing
from dataclasses import dataclass, field
from enum import StrEnum
from pathlib import Path

import structlog
from omegaconf import MISSING, DictConfig, MissingMandatoryValue, OmegaConf, ValidationError

log = structlog.get_logger()

PKG_INFO_FILE = Path("./common_packages.yaml")
UNKNOWN_VERSION = "UNKNOWN"
VERSION_RE = r"[vVr]?[0-9]+(\.[0-9]+)*"
# source: https://semver.org/#is-there-a-suggested-regular-expression-regex-to-check-a-semver-string
SEMVER_RE = r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"  # noqa: E501


class UpdatePolicy(StrEnum):
    AUTO = "Auto"
    PASSIVE = "Passive"


class PublishPolicy(StrEnum):
    HOMEASSISTANT = "HomeAssistant"
    MQTT = "MQTT"
    SILENT = "Silent"


class LogLevel(StrEnum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARNING = "WARNING"
    ERROR = "ERROR"
    CRITICAL = "CRITICAL"


class RegistryAPI(StrEnum):
    OCI_V2 = "OCI_V2"
    OCI_V2_MINIMAL = "OCI_V2"
    DOCKER_CLIENT = "DOCKER_CLIENT"
    DISABLED = "DISABLED"


class VersionType:
    SHORT_SHA = "short_sha"
    FULL_SHA = "full_sha"
    VERSION_REVISION = "version_revision"
    VERSION = "version"


@dataclass
class RegistryConfig:
    api: RegistryAPI = RegistryAPI.OCI_V2
    mutable_cache_ttl: int | None = None  # default to server cache hint
    immutable_cache_ttl: int | None = 7776000  # 90 days
    token_cache_ttl: int | None = None  # default to server cache hint


@dataclass
class MqttConfig:
    host: str = "${oc.env:MQTT_HOST,localhost}"
    user: str = f"${{oc.env:MQTT_USER,{MISSING}}}"
    password: str = f"${{oc.env:MQTT_PASS,{MISSING}}}"
    port: int = "${oc.decode:${oc.env:MQTT_PORT,1883}}"  # type: ignore[assignment]
    topic_root: str = "updates2mqtt"
    protocol: str = "${oc.env:MQTT_VERSION,3.11}"


@dataclass
class GitHubConfig:
    access_token: str | None = None


@dataclass
class MetadataSourceConfig:
    enabled: bool = True
    cache_ttl: int = 60 * 60 * 24 * 7  # 1 week


@dataclass
class Selector:
    include: list[str] | None = None
    exclude: list[str] | None = None


class VersionPolicy(StrEnum):
    AUTO = "AUTO"
    VERSION = "VERSION"
    DIGEST = "DIGEST"
    VERSION_DIGEST = "VERSION_DIGEST"
    TIMESTAMP = "TIMESTAMP"


@dataclass
class DockerPackageUpdateInfo:
    image_name: str = MISSING  # untagged image ref
    version_policy: VersionPolicy = VersionPolicy.AUTO


@dataclass
class PackageUpdateInfo:
    docker: DockerPackageUpdateInfo | None = field(default_factory=DockerPackageUpdateInfo)
    logo_url: str | None = None
    release_notes_url: str | None = None
    source_repo_url: str | None = None


@dataclass
class DockerConfig:
    enabled: bool = True
    allow_pull: bool = True
    allow_restart: bool = True
    allow_build: bool = True
    compose_version: str = "v2"
    default_entity_picture_url: str = "https://www.docker.com/wp-content/uploads/2022/03/Moby-logo.png"
    # Icon to show when browsing entities in Home Assistant
    device_icon: str = "mdi:docker"
    discover_metadata: dict[str, MetadataSourceConfig] = field(
        default_factory=lambda: {"linuxserver.io": MetadataSourceConfig(enabled=True)}
    )
    registry: RegistryConfig = field(default_factory=lambda: RegistryConfig())
    default_api_backoff: int = 60 * 15
    image_ref_select: Selector = field(default_factory=lambda: Selector())
    version_select: Selector = field(default_factory=lambda: Selector())
    version_policy: VersionPolicy = VersionPolicy.AUTO
    registry_select: Selector = field(default_factory=lambda: Selector())


@dataclass
class HomeAssistantDiscoveryConfig:
    prefix: str = "homeassistant"
    enabled: bool = True


@dataclass
class HomeAssistantConfig:
    discovery: HomeAssistantDiscoveryConfig = field(default_factory=HomeAssistantDiscoveryConfig)
    state_topic_suffix: str = "state"
    device_creation: bool = True
    force_command_topic: bool = False
    extra_attributes: bool = True
    area: str | None = None


@dataclass
class HealthCheckConfig:
    enabled: bool = True
    interval: int = 300  # Interval in seconds to publish healthcheck message, 0 to disable
    topic_template: str = "healthcheck/{node_name}/updates2mqtt"


@dataclass
class NodeConfig:
    name: str = field(default_factory=lambda: os.uname().nodename.replace(".local", ""))
    git_path: str = "/usr/bin/git"
    healthcheck: HealthCheckConfig = field(default_factory=HealthCheckConfig)


@dataclass
class LogConfig:
    level: LogLevel = "${oc.decode:${oc.env:U2M_LOG_LEVEL,INFO}}"  # type: ignore[assignment] # pyright: ignore[reportAssignmentType]


@dataclass
class Config:
    log: LogConfig = field(default_factory=LogConfig)  # pyright: ignore[reportArgumentType, reportCallIssue]
    node: NodeConfig = field(default_factory=NodeConfig)
    mqtt: MqttConfig = field(default_factory=MqttConfig)  # pyright: ignore[reportArgumentType, reportCallIssue]
    homeassistant: HomeAssistantConfig = field(default_factory=HomeAssistantConfig)
    docker: DockerConfig = field(default_factory=DockerConfig)
    github: GitHubConfig = field(default_factory=GitHubConfig)
    scan_interval: int = 60 * 60 * 3
    packages: dict[str, PackageUpdateInfo] = field(default_factory=dict)


@dataclass
class CommonPackages:
    common_packages: dict[str, PackageUpdateInfo] = field(default_factory=dict)


class IncompleteConfigException(BaseException):
    pass


def is_autogen_config() -> bool:
    env_var: str | None = os.environ.get("U2M_AUTOGEN_CONFIG")
    return not (env_var and env_var.lower() in ("no", "0", "false"))


def load_app_config(conf_file_path: Path, return_invalid: bool = False) -> Config | None:
    base_cfg: DictConfig = OmegaConf.structured(Config)
    if conf_file_path.exists():
        cfg: DictConfig = typing.cast("DictConfig", OmegaConf.merge(base_cfg, OmegaConf.load(conf_file_path)))
    elif is_autogen_config():
        if not conf_file_path.parent.exists():
            try:
                log.debug(f"Creating config directory {conf_file_path.parent} if not already present")
                conf_file_path.parent.mkdir(parents=True, exist_ok=True)
            except Exception as e:
                log.warning("Unable to create config directory: %s", e, path=conf_file_path.parent)
        try:
            conf_file_path.write_text(OmegaConf.to_yaml(base_cfg))
            log.info(f"Auto-generated a new config file at {conf_file_path}")
        except Exception as e:
            log.warning("Unable to write config file: %s", e, path=conf_file_path)
        cfg = base_cfg
    else:
        cfg = base_cfg

    try:
        # Validate that all required fields are present, throw exception now rather than when config first used
        OmegaConf.to_container(cfg, throw_on_missing=True)
        OmegaConf.set_readonly(cfg, True)
        config: Config = typing.cast("Config", cfg)

        if config.mqtt.user in ("", MISSING) or config.mqtt.password in ("", MISSING):
            log.info("The config has place holders for MQTT user and/or password")
            if not return_invalid:
                return None
        return config
    except (MissingMandatoryValue, ValidationError) as e:
        log.error("Configuration error %s", e, path=conf_file_path.as_posix())
        if return_invalid and cfg is not None:
            return typing.cast("Config", cfg)
        raise
